keps/sig-network/2091-cluster-network-policy/README.md

```diff
@@ -0,0 +1,1161 @@
+// added in the future to support other workload selection mechanism, e.g. serviceAccountSelector.
+// Must be accompanied by a NamespaceSelector to specify the scope of the workload selection.
+type WorkloadSelector struct {
+	PodSelector         *metav1.LabelSelector
+}
+```
+
+**Alternatives considered for the WorkloadSelector struct:**  
+```golang
+type WorkloadSelector struct {
+	selectorType        WorkloadSelectorType
+	selector            metav1.LabelSelector
+}
+
+type WorkloadSelectorType string
+
+// Add future selector types here to ease introduction of future selectors (e.g. serviceAccountSelector)
+// and ensure api fails closed cleanly if a CNI doesn't support specific types of selectors.
+const (
+	PodSelector WorkloadSelectorType  = "PodSelector"
+)
+```
+This alternative WorkloadSelector definition is more explicit but more verbose as well.
+At this point, it is expected that any future selectors will be mutually exclusive to the
+PodSelector in a single WorkloadSelector. Both approaches should be future extensible
+without breaking the premise of fail-closed API design.
+
+(TODO) Does there seem to be a valid usecase where two or more types of selectors will
+need to co-exist in a same workloadSelector? i.e. select workloads in these Namespaces,
+where it matches certain Pod labels, AND matches serviceAccount selection criteria?
+
+```golang
+type ClusterNetworkPolicyPeer struct {
+	WorkloadSelector    *WorkloadSelector
+	// One of NamespaceSelector or Namespaces is required.
+	// In the same ClusterNetworkPolicyPeer, NamespaceSelector and Namespaces fields are mutually
+	// exclusive.
+	NamespaceSelector   *metav1.LabelSelector
+	Namespaces          *Namespaces
+}
+
+const (
+	// RuleActionPass enables admins to provide exceptions to ClusterNetworkPolicies and delegate this rule to
+	// K8s NetworkPolicies.
+	RuleActionPass      RuleAction = "Pass"
+	// RuleActionDeny enables admins to deny specific traffic.
+	RuleActionDeny      RuleAction = "Deny"
+	// RuleActionAllow enables admins to specifically allow certain traffic.
+	RuleActionAllow     RuleAction = "Allow"
+)
+```
+
+For the ClusterNetworkPolicy ingress/egress rule, the `Action` field dictates whether
+traffic should be allowed or denied from/to the ClusterNetworkPolicyPeer. This will be a required field.
+
+The following structs will be added to the `netpol.networking.k8s.io` API group:
+
+```golang
+type AppliedTo struct {
+	// required if a PodSelector is specified
+	NamespaceSelector   *metav1.LabelSelector
+	// optional
+	WorkloadSelector    *WorkloadSelector
+}
+
+// Namespaces define a way to select Namespaces in the cluster.
+type Namespaces struct {
+	Scope       NamespaceMatchType
+	// Labels are set only when scope is "SameLabels".
+	Labels      []string
+	// Selector is only set when scope is "Selector".
+	// Namespaces.Selector has the same effect as NamespaceSelector.
+	Selector    *metav1.LabelSelector
```

Comment:
Why have `NamespaceSelector` at all then? Just get rid of it and leave the more expressive `Namespaces` field only.

("Because NetworkPolicy has a NamespaceSelector field" is not a good answer.)